### DISCUSSION TITLE
what is the difference？ zap  or   zap (sugared)

### DISCUSSION THREAD

**Discussion by icetech233** [uber-go/zap#1459] [General]
what is the difference？ zap  or   zap (sugared)

**Comment by adilburaksen** [comment]
Two log handles in zap, trading ergonomics for performance:

**`zap.Logger` (structured, typed)**
```go
logger, _ := zap.NewProduction()
logger.Info("server started",
    zap.String("host", "localhost"),
    zap.Int("port", 8080),
)
```
Fields must be typed explicitly. Zero allocations on the hot path. Use this in performance-critical code.

**`zap.SugaredLogger` (flexible, printf-style)**
```go
sugar := logger.Sugar()

// Printf-style
sugar.Infof("server started on %s:%d", host, port)

// Loosely-typed key-value pairs
sugar.Infow("server started", "host", "localhost", "port", 8080)
```
Accepts `interface{}` values — convenient but uses reflection, so ~1-2 allocations per call. Use this when log calls aren't in tight loops.

**When to use which:**

| | `Logger` | `SugaredLogger` |
|---|---|---|
| Performance | Zero alloc | ~1-2 allocs/call |
| Type safety | Compile-time | Runtime |
| Verbosity | More | Less |
| Use case | Hot paths, libraries | Application code, one-offs |

You can switch between them for free:
```go
sugar := logger.Sugar()    // Logger → SugaredLogger
logger = sugar.Desugar()   // SugaredLogger → Logger
```

The rule of thumb from the zap docs: use `Logger` in libraries and tight loops, `SugaredLogger` everywhere else.
